text/0000-match-ergonmics.md

```diff
@@ -0,0 +1,531 @@
+`j` by-value.
+
+
+### Safety
+
+Whenever considering relaxing Rust's rules around borrowing, we must consider
+safety. Could either of the proposed rules make Rust code less safe?
+
+Target auto-dereference introduces implicit dereferencing. However, there is no
+safety concern in the strict sense because the compiler still treats the data as
+borrowed and therefore, for example, won't let data move out of the `match`. We
+do not permit implicit dereferencing of owned data (e.g., `Box<_>`).
+
+With match arm auto-ref, there is no hard threat to safety - the compiler is
+merely inferring some annotations, the borrow checker remains unchanged and
+cannot be compromised.
+
+### Mental model
+
+One risk is that errors become even more confusing - the compiler might complain
+about borrowing violations without there being explicit dereferencing of the
+target expression. I believe that these errors are less likely due to arm auto-
+ref, and that when they do happen, there must be an explicit dereference in the
+arm body to indicate an error. Furthermore, since the target must always be
+borrowed, either its type or an explicit borrow (`&`) should indicate that the
+data is borrowed.
+
+It is reasonable to argue that arm auto-ref makes `match` more confusing
+(because there is an additional subtle rule to understand). I don't think this
+has proven to be a problem with closures, and I don't see why it should be with
+match arms, but some experience with the feature when implemented will help us
+decide.
+
+Note that explicit borrows where pattern variables are used are still required,
+as they would be if this feature did not exist. I.e., there is still a clear
+indication in the source code that owned data becomes borrowed.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The major downside of this proposal is that it introduces implicit referencing
+and dereferencing, and in particular implicit borrowing of otherwise un-borrowed
+data (in the arm auto-ref case). This is generally something we avoid, however,
+there is some precedent in the dot operator and closures.
+
+This is a fairly major change to the semantics of a common feature with no opt-
+in. Whilst I believe the proposal is backwards compatible (see above) and mostly
+only affects code which would not compile today, it is a bit unnerving.
+
+This RFC introduces differences in the semantics of patterns in `match` compared
+with `let`. On the other hand, there are differences already - allowing multiple
+patterns separated with `|` and `if` guards in match arms.
+
+As an alternative to this proposal, we could apply the same rules to `let` as to
+match arms. I think this should not be done because `ref` in `let` statements is
+much rarer than in matches, and there is no explicit scope to limit checking of
+uses to. These reasons make the facility less necessary and less appealing.
+
+
+# Alternatives
+[alternatives]: #alternatives
```

Comment:
One alternative is to not do any of these changes, but only improve the error messages and add suggestions for automatically applying the fixes that in this rfc are implicitly happening in the background. In fact,this could be one of the cases where the compiler could set the hypothetical confidence flag as suggested in #1941